src/coreclr/src/jit/morph.cpp

```diff
@@ -12515,25 +12515,49 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 op2 = tree->AsOp()->gtOp2;
             }
 
-            /* See if we can fold GT_ADD nodes. */
-
-            if (oper == GT_ADD)
+            // See if we can fold constants for commutative operators.
+            if (tree->OperIs(GT_OR, GT_XOR, GT_AND, GT_ADD))
             {
-                /* Fold "((x+icon1)+(y+icon2)) to ((x+y)+(icon1+icon2))" */
-
-                if (op1->gtOper == GT_ADD && op2->gtOper == GT_ADD && !gtIsActiveCSE_Candidate(op2) &&
-                    op1->AsOp()->gtOp2->gtOper == GT_CNS_INT && op2->AsOp()->gtOp2->gtOper == GT_CNS_INT &&
-                    !op1->gtOverflow() && !op2->gtOverflow())
+                // Fold "((x <op> icon1) <op> (y <op> icon2))" to "((x <op> y) <op> (icon1 <op> icon2))"
+                if (op1->OperIs(op2->OperGet()) && op1->OperIs(oper) && !gtIsActiveCSE_Candidate(op2) &&
+                    op1->AsOp()->gtGetOp2()->IsCnsIntOrI() && op2->AsOp()->gtGetOp2()->IsCnsIntOrI())
                 {
                     // Don't create a byref pointer that may point outside of the ref object.
                     // If a GC happens, the byref won't get updated. This can happen if one
                     // of the int components is negative. It also requires the address generation
                     // be in a fully-interruptible code region.
-                    if (!varTypeIsGC(op1->AsOp()->gtOp1->TypeGet()) && !varTypeIsGC(op2->AsOp()->gtOp1->TypeGet()))
+                    if (!varTypeIsGC(op1->AsOp()->gtGetOp1()->TypeGet()) &&
+                        !varTypeIsGC(op2->AsOp()->gtGetOp1()->TypeGet()))
                     {
-                        cns1 = op1->AsOp()->gtOp2;
-                        cns2 = op2->AsOp()->gtOp2;
-                        cns1->AsIntCon()->gtIconVal += cns2->AsIntCon()->gtIconVal;
+                        cns1          = op1->AsOp()->gtGetOp2();
+                        cns2          = op2->AsOp()->gtGetOp2();
+                        ssize_t icon1 = cns1->AsIntCon()->IconValue();
+                        ssize_t icon2 = cns2->AsIntCon()->IconValue();
+
+                        if (oper == GT_ADD)
+                        {
+                            if (op1->gtOverflow() || op2->gtOverflow())
```

Comment:
I understand that that is pre-existing, but don't we need to check (tree->gtOverflow()) as well?

With the current implementation, we could convert
`((x+(INT_MAX/2 + 1))+(y+(INT_MAX/2 + 1)))` to `((x+y)+((INT_MAX/2 + 1)+(INT_MAX/2 + 1)))`, couldn't we?

And then we will call `12565: cns1->AsIntCon()->TruncateOrSignExtend32();` for `((INT_MAX/2 + 1)+(INT_MAX/2 + 1))`  so it would hide the overflow and change `INT_MAX + 2` to `2`, is that correct?

Also, comment on 12564 should be updated, because now it is not only `adding`.

Reply:
tree->gtOverflow is handled [here](https://github.com/dotnet/runtime/blob/afc21e51af60a4d2cb11a1081fb38b1e3a8df0a2/src/coreclr/src/jit/morph.cpp#L12489).
but I added it to this condition just in case too.